library/n2_protocol.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef ELLIPTICS_N2_PROTOCOL_HPP
+#define ELLIPTICS_N2_PROTOCOL_HPP
+
+#include <optional>
+
+#include <elliptics/packet.h>
+#include <elliptics/utils.hpp>
+
+namespace ell = ioremap::elliptics;
+
+namespace ioremap { namespace elliptics { namespace n2 {
+
+using cmd_native = dnet_cmd;
+
+struct data_in_file {
+	int fd = -1;
+	off_t local_offset = 0;
+	size_t fsize = 0;
+	int on_exit = 0;
+};
+
+struct data_place {
+	enum place_type {
+		IN_MEMORY,
+		IN_FILE,
+	};
+
+	data_in_file in_file;
+	ell::data_pointer in_memory;
+
+	place_type where() const;
+
+	static data_place from_file(const data_in_file &in_file);
+	static data_place from_memory(const ell::data_pointer &in_memory);
+};
+
+class isender;
+
+}}} // namespace ioremap::elliptics::n2
+
+// All requests/responses must be inherited from common_request;
+// common_request is visible from dnet_io_req. Since this struct
+// is visible from C code, it is declared out of namespace.
+struct n2_message {
+	ell::n2::cmd_native cmd;
+	std::optional<ell::n2::data_place> data;
```

Comment:
Why is it part of `n2_message`?

Reply:
To support logic of `dnet_io_req_copy`. If we use `n2_message` both for old and new protocol, we should, at least, to support two alternatives of data (`fd` or `data pointer`) that presented in old protocol: there was common logic, independent on request type.

Reply:
Alternative: in each message type data behavior is specific, conversion (if necessary) is done via call nonconst virtual method `n2_message::make_owning()`. This method by default will do nothing, but in messages that use fd this method will read data from file.